cpp/src/arrow/compute/exec/union_node_test.cc

```diff
@@ -17,187 +17,126 @@
+                                 .AddToPlan(plan.get()));
+      return;
+    } else {
+      ASSERT_OK(Declaration::Sequence({union_decl, {"sink", SinkNodeOptions{&sink_gen}}})
+                    .AddToPlan(plan.get()));
+    }
 
-  CheckRunOutput(l_empty, r_empty, output_batches);
-}
+    Future<std::vector<ExecBatch>> actual = StartAndCollect(plan.get(), sink_gen);
 
-TEST(UnionTest, TestNonEmpty) {
-  for (bool parallel : {false, true}) {
-    RunNonEmptyTest(parallel);
+    auto expected_matcher =
+        Finishes(ResultWith(UnorderedElementsAreArray(exp_batches.batches)));
+    ASSERT_THAT(actual, expected_matcher);
   }
-}
 
-TEST(UnionTest, TestEmpty) {
-  for (bool parallel : {false, true}) {
-    RunEmptyTest(parallel);
+  void CheckUnionExecNode(size_t num_input_nodes, size_t num_batches, bool parallel) {
+    auto random_schema = GenerateRandomSchema(num_input_nodes);
+
+    int multiplicity = parallel ? 10 : 1;
+    std::vector<std::shared_ptr<RecordBatch>> all_record_batches;
+    std::vector<BatchesWithSchema> input_batches(num_input_nodes);
+    BatchesWithSchema exp_batches;
+    exp_batches.schema = random_schema;
+    for (size_t i = 0; i < num_input_nodes; i++) {
+      GenerateBatchesFromSchema(random_schema, num_batches, &input_batches[i],
+                                multiplicity, kBatchSize);
+      for (const auto& batch : input_batches[i].batches) {
+        exp_batches.batches.push_back(batch);
+      }
+    }
+    CheckRunOutput(input_batches, exp_batches, parallel);
   }
-}
-
-void TestUnionRandom(const std::shared_ptr<DataType>& data_type, bool parallel,
-                     int num_batches, int batch_size) {
-  auto l_schema = schema({field("colum0", data_type), field("colum1", data_type)});
-  auto r_schema = schema({field("colum0", data_type), field("colum1", data_type)});
-
-  // generate data
-  auto l_batches = MakeRandomBatches(l_schema, num_batches, batch_size);
-  auto r_batches = MakeRandomBatches(r_schema, num_batches, batch_size);
-
-  ASSERT_OK_AND_ASSIGN(auto plan, ExecPlan::Make());
-
-  Declaration Union{"union", ExecNodeOptions{}};
-
-  // add left source
-  Union.inputs.emplace_back(Declaration{
-      "source", SourceNodeOptions{l_batches.schema, l_batches.gen(parallel,
-                                                                  /*slow=*/false)}});
-  // add right source
-  Union.inputs.emplace_back(Declaration{
-      "source", SourceNodeOptions{r_batches.schema, r_batches.gen(parallel,
-                                                                  /*slow=*/false)}});
-  AsyncGenerator<util::optional<ExecBatch>> sink_gen;
 
-  ASSERT_OK(Declaration::Sequence({Union, {"sink", SinkNodeOptions{&sink_gen}}})
-                .AddToPlan(plan.get()));
+  ::arrow::random::SeedType seed_ = 0xdeadbeef;
+  ::arrow::random::RandomArrayGenerator rng_;
 
-  auto actual = StartAndCollect(plan.get(), sink_gen);
+  static constexpr int kNumBatches = 10;
```

Comment:
nit: put constants towards the top. See https://google.github.io/styleguide/cppguide.html#Declaration_Order